src/Analyser/TypeSpecifier.php

```diff
@@ -1034,6 +896,215 @@ public function specifyTypesInCondition(
+					$funcTypes = $this->create($exprNode, $constantType, $context, false, $scope, $rootExpr);
+
+					$accessory = new AccessoryNonEmptyStringType();
+					if ($constantType->getValue() >= 2) {
+						$accessory = new AccessoryNonFalsyStringType();
+					}
+					$valueTypes = $this->create($exprNode->getArgs()[0]->value, $accessory, $newContext, false, $scope, $rootExpr);
+
+					return $funcTypes->unionWith($valueTypes);
+				}
+			}
+
+		}
+
+		if ($constantType instanceof ConstantStringType) {
+			return $this->specifyTypesForConstantStringBinaryExpression($exprNode, $constantType, $context, $scope, $rootExpr);
+		}
+
+		return null;
+	}
+
+	private function specifyTypesForConstantStringBinaryExpression(
+		Expr $exprNode,
+		ConstantStringType $constantType,
+		TypeSpecifierContext $context,
+		Scope $scope,
+		?Expr $rootExpr,
+	): ?SpecifiedTypes
+	{
+		if (
+			$context->truthy()
+			&& $exprNode instanceof FuncCall
+			&& $exprNode->name instanceof Name
+			&& in_array(strtolower($exprNode->name->toString()), ['substr', 'strstr', 'stristr', 'strchr', 'strrchr', 'strtolower', 'strtoupper', 'mb_strtolower', 'mb_strtoupper', 'ucfirst', 'lcfirst', 'ucwords'], true)
+			&& isset($exprNode->getArgs()[0])
+			&& $constantType->getValue() !== ''
+		) {
+			$argType = $scope->getType($exprNode->getArgs()[0]->value);
+
+			if ($argType->isString()->yes()) {
+				if ($constantType->getValue() !== '0') {
+					return $this->create(
+						$exprNode->getArgs()[0]->value,
+						TypeCombinator::intersect($argType, new AccessoryNonFalsyStringType()),
+						$context,
+						false,
+						$scope,
+					);
+				}
+
+				return $this->create(
+					$exprNode->getArgs()[0]->value,
+					TypeCombinator::intersect($argType, new AccessoryNonEmptyStringType()),
+					$context,
+					false,
+					$scope,
+				);
+			}
+		}
+
+		if (
+			$exprNode instanceof FuncCall
+			&& $exprNode->name instanceof Name
+			&& strtolower($exprNode->name->toString()) === 'gettype'
+			&& isset($exprNode->getArgs()[0])
+		) {
+			$type = null;
+			if ($constantType->getValue() === 'string') {
+				$type = new StringType();
+			}
+			if ($constantType->getValue() === 'array') {
+				$type = new ArrayType(new MixedType(), new MixedType());
+			}
+			if ($constantType->getValue() === 'boolean') {
+				$type = new BooleanType();
+			}
+			if ($constantType->getValue() === 'resource' || $constantType->getValue() === 'resource (closed)') {
+				$type = new ResourceType();
+			}
+			if ($constantType->getValue() === 'integer') {
+				$type = new IntegerType();
+			}
+			if ($constantType->getValue() === 'double') {
+				$type = new FloatType();
+			}
+			if ($constantType->getValue() === 'NULL') {
+				$type = new NullType();
+			}
+			if ($constantType->getValue() === 'object') {
+				$type = new ObjectWithoutClassType();
+			}
+
+			if ($type !== null) {
+				return $this->create($exprNode->getArgs()[0]->value, $type, $context, false, $scope, $rootExpr);
+			}
+		}
+
+		if (
```

Comment:
this if block is the only new logic of this PR.

everything else is pre-existing logic which was just factored out into new private methods